GCPSigner import

jku opened the pull request:
Implement import  (loading the whole key, including public key, from the KMS) for GCPSigner.

This is the implementation of #466 for this specific Signer -- in my opinion this can be merged: if we decide to change something in #466 that affects this Signer we can do that later.

At this point GCPSigner is fully usable: I don't think it needs anything more.

### Please verify and check that the pull request fulfils the following requirements:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

jku commented:
test run in https://github.com/jku/securesystemslib/actions/runs/3677912828/jobs/6220505707

jku commented:
Marking ready for review

jku commented:
I'll add a couple of commits that try to "standardize" GCP and HSM "key creation / import": these change HSMSigner mostly but I think it makes some sense as part of this PR: the goal is to keep the offered Signer APIs coherent, even if there is no specific signature that the signers are implementing.

* use the same method name `import_()`
* same return values: priv key uri and public key
* and finally, handle keyid creation internally

The last one, keyid change,  is a separate commit so it can be dropped if you don't like it. My reasons for including it: 
* It's true that callers should have the opportunity to set keyids, but I think they do have that: they can just set the keyid of the key to whatever they want after getting it from import_()
* providing a reasonable default still seems nice (the implementation may not be pretty but now it's an implementation detail that we can change without API changing)

Neither signer has been released yet so API changes are fine.

lukpueh commented on securesystemslib/signer/_gcp_signer.py:
```diff
@@ -71,6 +84,79 @@ def from_priv_key_uri(
+        """Return keytype and scheme for the KMS algorithm enum"""
+        keytypes_and_schemes = {
+            CryptoKeyVersion.CryptoKeyVersionAlgorithm.EC_SIGN_P256_SHA256: (
+                "ecdsa-sha2-nistp256",
```
Minor nit: https://github.com/secure-systems-lab/securesystemslib/issues/363 suggests to deprecate `ecdsa-sha2-nistp256` as keytype in favor of the more succinct `ecdsa`.

jku replied:
oh, I must have gotten these mixed up

lukpueh commented on securesystemslib/signer/_gcp_signer.py:
```diff
@@ -71,6 +84,79 @@ def from_priv_key_uri(
+                "rsa-pkcs1v15-sha512",
+            ),
+        }
+        return keytypes_and_schemes[algorithm]
```
Is it worth creating a ticket to test these all schemes?

jku replied:
I suppose that would be nice -- it will require creating all of these keys in the KMS though

lukpueh reviewed (commented):
LGTM. Please comment on `priv_key_uri` signer attribute and I'll approve. Other inline comments are non-blockers.

jku commented on securesystemslib/signer/_gcp_signer.py:
```diff
@@ -54,6 +66,7 @@ def __init__(self, gcp_keyid: str, public_key: Key):
 
         self.hash_algorithm = self._get_hash_algorithm(public_key)
         self.gcp_keyid = gcp_keyid
+        self.priv_key_uri = f"{self.SCHEME}:{gcp_keyid}"
```
I think there is no benefit -- I added that before we arrived at the conclusion that it makes sense as import return value. I will  remove the attribute

jku commented:
interesting, I broke the kms test. Investigating...

lukpueh commented:
[Test still failing](https://github.com/jku/securesystemslib/actions/runs/3894425830/jobs/6648399377). Could it be that you have to recalculate the keyid? 

https://github.com/secure-systems-lab/securesystemslib/blob/8f85dc8c84ad8b57641209ab556c5cb63751afa0/tests/check_kms_signers.py#L28

lukpueh commented on tests/check_kms_signers.py:
```diff
@@ -18,13 +18,25 @@
 
-    def test_gcp(self):
+    pubkey = Key.from_dict(
+        "23b650fcc538c0c6d423d0886448172fd51d81f43ef749851a0a84978cdcd8e0",
```
```suggestion
        "218611b80052667026c221f8774249b0f6b8b310d30a5c45a3b878aa3a02f39e",
```
should work

jku commented:
Okay: the test was failing as it should since I changed the key content. I fixed the test content to match as suggested, test now passes: 

https://github.com/jku/securesystemslib/actions/runs/3894855218/jobs/6649416664
